crates/matrix-sdk-crypto/src/machine/mod.rs

```diff
@@ -1336,7 +1359,54 @@ impl OlmMachine {
             e => self.handle_to_device_event(changes, &e).await,
         }
 
-        raw_event
+        Some(raw_event)
+    }
+
+    async fn event_is_from_dehydrated_device(
+        &self,
+        decrypted: &OlmDecryptionInfo,
+        sender_user_id: &UserId,
+    ) -> bool {
+        // Does the to-device message include device info?
+        if let Some(device_keys) = decrypted.result.event.sender_device_keys() {
+            // There is no need to check whether the device keys are signed correctly - any
+            // to-device message that claims to be from a dehydrated device is weird, so we
+            // will drop it.
+
+            // Does the included device info say the device is dehydrated?
+            if device_keys.dehydrated.unwrap_or(false) {
+                return true;
+            }
+            // If not, fall through and check our existing list of devices
+            // below, just in case the sender is sending us
+            // incorrect information embedded in the to-device message, but we
+            // know better.
+        }
+
+        // Do we already know about this device?
+        match self
+            .store()
+            .get_device_from_curve_key(sender_user_id, decrypted.result.sender_key)
+            .await
+        {
+            Ok(device) => {
+                // Yes - is it dehydrated?
+                if let Some(device) = device {
+                    device.is_dehydrated()
+                } else {
+                    false
+                }
+            }
+            Err(e) => {
+                // No - we received a to-device message from a device we don't know about. We
+                // continue to process it, but this is unexpected so we log an error.
+                warn!(
+                    error = ?e,
+                    "Received a to-device message from a device we don't recognise",
+                );
```

Comment:
I think if we just don't recognise the device, `get_device_from_curve_key` will return `Ok(None)`. This suggests a problem with the store.

It's probably still ok to return `false` here; but the comment and warning could do with a tweak.

Reply:
Good catch, thanks. I surfaced the error in dd7bfeaf73 , following the pattern of nearby code about what to do with a storage error.